File: 프로그래머스/3/43162. 네트워크/네트워크.js
class UnionFind {
    constructor(n){
        this.parent = Array.from({length: n}, (_, i)=>i);
        this.rank = new Array(n).fill(1);
    }
    
    find(x){
        if (this.parent[x]!==x){
            this.parent[x]=this.find(this.parent[x]);
        }
        return this.parent[x];
    }
    
    union(x, y){
        let rootX = this.find(x);
        let rootY = this.find(y);
        if (rootX!==rootY){
            if (this.rank[rootX]>this.rank[rootY]){
                this.parent[rootY] = rootX;
            } else if (this.rank[rootX]<this.rank[rootY]){
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX;
                this.rank[rootX]++;
            }
        }
    }
}

function solution(n, computers) {
    const uf = new UnionFind(n);
    for (let r=0; r<computers.length; r++){
        for (let c=0; c<computers[0].length; c++){
            if (computers[r][c]===1){
                uf.union(r, c);
            }
        }
    }
    const set = new Set();
    for (let i=0; i<uf.parent.length; i++){
        set.add(uf.find(uf.parent[i]));
    }
    return set.size;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    static class UnionFind {
        int[] parent, rank;
        UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i=0; i<n; i++) {
                parent[i] = i;
                rank[i] = 1;
            }
        }
        
        int find(int x) {
            if (parent[x]!=x){
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }
        
        void union(int a, int b) {
            int rootA = find(a);
            int rootB = find(b);
            
            if (rank[rootA]>rank[rootB]) parent[rootB] = rootA;
            else if (rank[rootA]<rank[rootB]) parent[rootA] = rootB;
            else {
                rank[rootA]++;
                parent[rootB] = rootA;
            }
        }
    }
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        UnionFind uf = new UnionFind(n);
        
        for (int r=0; r<n; r++){
            for (int c=0; c<n; c++) {
                if (computers[r][c]==1) uf.union(r, c);
            }
        }
        
        Set<Integer> set = new HashSet<>();
        for (int i=0; i<n; i++) {
            set.add(uf.find(i));
        }
        
        return set.size();
    }
}